Putting HashMap in a heap block hurts memory use, speed, code size (part 1)

darinadler opened the pull request:
#### 0c95ee27a86ea97629da9bc1dfe216be3317814c
<pre>
Putting HashMap in a heap block hurts memory use, speed, code size (part 1)
<a href="https://bugs.webkit.org/show_bug.cgi?id=226162">https://bugs.webkit.org/show_bug.cgi?id=226162</a>
<a href="https://rdar.apple.com/78675956">rdar://78675956</a>

Reviewed by Keith Miller.

HashMap, HashSet, and HashCountedSet are already pointers to a hash table,
so there is no reason to wrap them in a unique_ptr in a heap block and add
another level of indirection. Fix this in a few places in JavaScriptCore.

* Source/JavaScriptCore/API/JSBase.cpp:
(JSGetMemoryUsageStatistics): Removed pointer dereference.
* Source/JavaScriptCore/API/JSClassRef.cpp:
(OpaqueJSClass::~OpaqueJSClass): Use simpler for loops.
(OpaqueJSClassContextData::OpaqueJSClassContextData): Ditto.
(OpaqueJSClass::staticValues): No need to call unique_ptr::get().
(OpaqueJSClass::staticFunctions): Ditto.
* Source/JavaScriptCore/API/JSClassRef.h: Stop using unique_ptr for
the static values and functions tables.

* Source/JavaScriptCore/ftl/FTLAbstractHeap.cpp:
(JSC::FTL::IndexedAbstractHeap::atSlow): Removed code to allocate m_largeIndices.
* Source/JavaScriptCore/ftl/FTLAbstractHeap.h: Made m_largeIndices a map, not
a unique_ptr to a map.

* Source/JavaScriptCore/heap/CompleteSubspace.cpp:
(JSC::CompleteSubspace::reallocatePreciseAllocationNonVirtual): Update since
preciseAllocationSet is now an optional rather than a pointer to a heap block.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::protectedObjectTypeCounts): Return a set instead of a pointer holding one.
(JSC::Heap::objectTypeCounts): Ditto.
* Source/JavaScriptCore/heap/Heap.h: Ditto.

* Source/JavaScriptCore/heap/HeapUtil.h:
(JSC::HeapUtil::isPointerGCObjectJSCell): Update since preciseAllocationSet is now an
optional rather than a pointer to a heap block. Also remove an unneeded empty check
since the contains function already handles that case so it&apos;s just redundant.

* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::registerPreciseAllocation): Update since preciseAllocationSet is
now an optional rather than a pointer to a heap block.
(JSC::MarkedSpace::sweepPreciseAllocations): Ditto.
(JSC::MarkedSpace::enablePreciseAllocationTracking): Ditto.

* Source/JavaScriptCore/heap/MarkedSpace.h:
(JSC::MarkedSpace::preciseAllocationSet): Use std::optional instead of std::unique_ptr.

* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::logMemoryStatistics): Update since objectTypeCounts now returns a set
instead of a unique_ptr to a set.
* Source/WebCore/page/PerformanceLogging.cpp:
(WebCore::PerformanceLogging::javaScriptObjectCounts): Ditto.
* Source/WebKitLegacy/mac/Misc/WebCoreStatistics.mm:
(+[WebCoreStatistics javaScriptProtectedObjectTypeCounts]): Ditto.
(+[WebCoreStatistics javaScriptObjectTypeCounts]): Ditto.

Canonical link: <a href="https://commits.webkit.org/297725@main">https://commits.webkit.org/297725@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5612da79e061c3a99a00ae3d7d93f4b9dd4d8da6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111350 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31016 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21481 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117382 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61618 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113312 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31697 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39598 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84633 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114297 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25311 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100245 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65080 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24654 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18383 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61202 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103843 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94695 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18450 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120558 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109904 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38399 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28534 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93559 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38775 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96517 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93383 "Found 2 new API test failures: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24086 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38484 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16257 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34413 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38288 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43765 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/134179 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37953 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36172 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41286 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39655 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/ProgramExecutable.cpp:
```diff
@@ -287,9 +287,9 @@ void ProgramExecutable::visitChildrenImpl(JSCell* cell, Visitor& visitor)
     ASSERT_GC_OBJECT_INHERITS(thisObject, info());
     Base::visitChildren(thisObject, visitor);
-    if (TemplateObjectMap* map = thisObject->m_templateObjectMap.get()) {
+    if (auto& map = thisObject->m_templateObjectMap; !map.isEmpty()) {
```
Is this thread safe? If the JS mutator is resizing this right as the marker thread looks at the map this could UAF. I think this should be `!map.isNullStorage()`. 

That said it's a bit unclear if this optimization is valid to begin with. It might be valid because adding something to the map does a `writeBarrier` on the executable.

darinadler replied:
I'll undo this part; I think we could use a thread-safe boolean for this instead of checking the status of the hash table with a tricky use of a write barrier, but for now I'll just roll this part back to the old way.

kmiller68 commented on Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
```diff
@@ -99,9 +99,9 @@ void ModuleProgramExecutable::visitChildrenImpl(JSCell* cell, Visitor& visitor)
     Base::visitChildren(thisObject, visitor);
     visitor.append(thisObject->m_moduleEnvironmentSymbolTable);
-    if (TemplateObjectMap* map = thisObject->m_templateObjectMap.get()) {
+    if (auto& map = thisObject->m_templateObjectMap; !map.isEmpty()) {
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
```diff
@@ -86,9 +86,9 @@ void FunctionExecutable::visitChildrenImpl(JSCell* cell, Visitor& visitor)
         visitor.append(rareData->m_cachedPolyProtoStructureID);
         visitor.append(rareData->m_asString);
-        if (TemplateObjectMap* map = rareData->m_templateObjectMap.get()) {
+        if (auto& map = rareData->m_templateObjectMap; !map.isEmpty()) {
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/runtime/EvalExecutable.cpp:
```diff
@@ -56,9 +56,9 @@ void EvalExecutable::visitChildrenImpl(JSCell* cell, Visitor& visitor)
     ASSERT_GC_OBJECT_INHERITS(thisObject, info());
     Base::visitChildren(thisObject, visitor);
-    if (TemplateObjectMap* map = thisObject->m_templateObjectMap.get()) {
+    if (auto& map = thisObject->m_templateObjectMap; !map.isEmpty()) {
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/heap/HeapUtil.h:
```diff
@@ -46,13 +46,13 @@ class HeapUtil {
             ASSERT(set);
-            if (set->isEmpty())
-                return false;
 #if USE(JSVALUE32_64)
             // In 32bit systems a cell pointer can be 0xFFFFFFFF (an entries in the call frame), and this
             // value clashes with the deletedValue in a set<JSCell*>.
+            // FIXME: In this case, the ASSERT(set) above could fail too so this check is too late.
+            // FIXME: Why couldn't a cell pointer be 0xFFFFFFFFFFFFFFFF on 64-bit systems too?
             if (!set->isValidValue(pointer))
                 return false;
 #endif
```
This all seems somewhat unnecessary for a number of reasons. Cell allocations are 16-byte aligned (shifted by 8 bytes for a precise allocation). So the `isPreciseAllocation()`  check would eliminate `0xFFFFFFFF` (and `0xFFFFFFFFFFFFFFFF` on 64-bit). Also, JSC strongly relies on a 48-bit pointers on 64-bit platforms so it wouldn't be `0xFFFFFFFFFFFFFFFF`. It seems like the callers of this function should pre-filter out things that have impossible bits set.

darinadler replied:
Test case that showed this was needed in 32-bit is in 55fce2d2fa0fc5c39d9f7d198eb3ff91cbd03c3b.

If it's better to put a check at the caller that's fine with me and if you want me to code it I will. If you don't mind, I will just leave it for now.

darinadler replied:
These values are garbage on the stack, presumably. Not actual pointers. I agree that we could do other types of checks to weed out bad values, but not **sure** they'd be significantly better.

kmiller68 commented on Source/JavaScriptCore/heap/MarkedSpace.h:
```diff
@@ -207,7 +207,7 @@ class MarkedSpace {
     Vector<Subspace*> m_subspaces;
 
-    std::unique_ptr<UncheckedKeyHashSet<HeapCell*>> m_preciseAllocationSet;
+    std::optional<UncheckedKeyHashSet<HeapCell*>> m_preciseAllocationSet;
```
Is there a reason this is an optional at all? You can't have a JSGlobalObject without allocating at least a few precise allocations so this should effectively always be initialized.

darinadler replied:
This set isn't created lazily when the first precise allocation is done. It's created if and when `enablePreciseAllocationTracking` is called. I don't know what the rationale is for having both modes.

kmiller68 reviewed (approved):
r=me with comments

darinadler committed:
Putting HashMap in a heap block hurts memory use, speed, code size (part 1)
https://bugs.webkit.org/show_bug.cgi?id=226162
rdar://78675956

Reviewed by Keith Miller.

HashMap, HashSet, and HashCountedSet are already pointers to a hash table,
so there is no reason to wrap them in a unique_ptr in a heap block and add
another level of indirection. Fix this in a few places in JavaScriptCore.

* Source/JavaScriptCore/API/JSBase.cpp:
(JSGetMemoryUsageStatistics): Removed pointer dereference.
* Source/JavaScriptCore/API/JSClassRef.cpp:
(OpaqueJSClass::~OpaqueJSClass): Use simpler for loops.
(OpaqueJSClassContextData::OpaqueJSClassContextData): Ditto.
(OpaqueJSClass::staticValues): No need to call unique_ptr::get().
(OpaqueJSClass::staticFunctions): Ditto.
* Source/JavaScriptCore/API/JSClassRef.h: Stop using unique_ptr for
the static values and functions tables.

* Source/JavaScriptCore/ftl/FTLAbstractHeap.cpp:
(JSC::FTL::IndexedAbstractHeap::atSlow): Removed code to allocate m_largeIndices.
* Source/JavaScriptCore/ftl/FTLAbstractHeap.h: Made m_largeIndices a map, not
a unique_ptr to a map.

* Source/JavaScriptCore/heap/CompleteSubspace.cpp:
(JSC::CompleteSubspace::reallocatePreciseAllocationNonVirtual): Update since
preciseAllocationSet is now an optional rather than a pointer to a heap block.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::protectedObjectTypeCounts): Return a set instead of a pointer holding one.
(JSC::Heap::objectTypeCounts): Ditto.
* Source/JavaScriptCore/heap/Heap.h: Ditto.

* Source/JavaScriptCore/heap/HeapUtil.h:
(JSC::HeapUtil::isPointerGCObjectJSCell): Update since preciseAllocationSet is now an
optional rather than a pointer to a heap block. Also remove an unneeded empty check
since the contains function already handles that case so it's just redundant.

* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::registerPreciseAllocation): Update since preciseAllocationSet is
now an optional rather than a pointer to a heap block.
(JSC::MarkedSpace::sweepPreciseAllocations): Ditto.
(JSC::MarkedSpace::enablePreciseAllocationTracking): Ditto.

* Source/JavaScriptCore/heap/MarkedSpace.h:
(JSC::MarkedSpace::preciseAllocationSet): Use std::optional instead of std::unique_ptr.

* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::logMemoryStatistics): Update since objectTypeCounts now returns a set
instead of a unique_ptr to a set.
* Source/WebCore/page/PerformanceLogging.cpp:
(WebCore::PerformanceLogging::javaScriptObjectCounts): Ditto.
* Source/WebKitLegacy/mac/Misc/WebCoreStatistics.mm:
(+[WebCoreStatistics javaScriptProtectedObjectTypeCounts]): Ditto.
(+[WebCoreStatistics javaScriptObjectTypeCounts]): Ditto.

Canonical link: https://commits.webkit.org/297725@main
